File: apps/mobile/native/android/ToxicBufferModule.kt
package ai.sibyl.ghost

import android.util.Base64
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import java.util.concurrent.ConcurrentHashMap

class ToxicBufferModule(private val context: ReactApplicationContext) :
  ReactContextBaseJavaModule(context) {

  private val buffers = ConcurrentHashMap<String, ByteArray>()

  override fun getName(): String = "ToxicBufferModule"

  @ReactMethod
  fun allocFromBase64(input: String, promise: Promise) {
    val handle = "tox_" + System.nanoTime().toString()
    val data = Base64.decode(input, Base64.DEFAULT)
    buffers[handle] = data
    promise.resolve(handle)
  }

  @ReactMethod
  fun readAsBase64(handle: String, promise: Promise) {
    val data = buffers[handle]
    if (data == null) {
      promise.reject("NOT_FOUND", "Buffer handle not found")
      return
    }
    promise.resolve(Base64.encodeToString(data, Base64.NO_WRAP))
  }

  @ReactMethod
  fun zeroizeAndFree(handle: String, promise: Promise) {
    val data = buffers.remove(handle)
    if (data != null) {
      for (i in data.indices) {
        data[i] = 0
      }
    }
    promise.resolve(null)
  }
}

File: apps/mobile/native/ios/ToxicBufferModule.swift
import Foundation
import React

@objc(ToxicBufferModule)
class ToxicBufferModule: NSObject {
  private var buffers: [String: Data] = [:]

  @objc
  func allocFromBase64(_ input: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
    guard let data = Data(base64Encoded: input) else {
      reject("INVALID_BASE64", "Could not decode base64", nil)
      return
    }

    let handle = "tox_\(UUID().uuidString)"
    buffers[handle] = data
    resolve(handle)
  }

  @objc
  func readAsBase64(_ handle: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
    guard let data = buffers[handle] else {
      reject("NOT_FOUND", "Buffer handle not found", nil)
      return
    }

    resolve(data.base64EncodedString())
  }

  @objc
  func zeroizeAndFree(_ handle: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
    if var data = buffers.removeValue(forKey: handle) {
      data.resetBytes(in: 0..<data.count)
    }
    resolve(nil)
  }

  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }
}
